File: go/capstone-message-queue/concurrent/concurrent_test.go
package concurrent_test

import (
	"sync"
	"testing"
	"time"

	"github.com/andersonreyes/learning/go/capstone-message-queue/broker"
	"github.com/andersonreyes/learning/go/capstone-message-queue/concurrent"
)

// ── helpers ───────────────────────────────────────────────────────────────────

func makeShared(t *testing.T, topics []struct{ name string; parts uint32 }) *concurrent.SharedRegistry {
	t.Helper()
	dir := t.TempDir()
	reg, err := broker.Open(dir)
	if err != nil {
		t.Fatalf("Open: %v", err)
	}
	for _, tp := range topics {
		if err := reg.CreateTopic(tp.name, tp.parts); err != nil {
			t.Fatalf("CreateTopic: %v", err)
		}
	}
	sr := concurrent.New(reg, 50*time.Millisecond)
	t.Cleanup(sr.Close)
	return sr
}

func totalOffsets(t *testing.T, sr *concurrent.SharedRegistry, topic string, numPartitions uint32) uint64 {
	t.Helper()
	var total uint64
	for p := uint32(0); p < numPartitions; p++ {
		off, err := sr.NextOffset(topic, p)
		if err != nil {
			t.Fatalf("NextOffset(%s, %d): %v", topic, p, err)
		}
		total += off
	}
	return total
}

// ── single-threaded API sanity ─────────────────────────────────────────────────

func TestCreateTopicAndProduceFetch(t *testing.T) {
	dir := t.TempDir()
	reg, _ := broker.Open(dir)
	sr := concurrent.New(reg, 100*time.Millisecond)
	defer sr.Close()

	if err := sr.CreateTopic("greetings", 1); err != nil {
		t.Fatalf("CreateTopic: %v", err)
	}
	pid, off, err := sr.Produce("greetings", []byte("hello"), nil)
	if err != nil {
		t.Fatalf("Produce: %v", err)
	}
	if pid != 0 || off != 0 {
		t.Fatalf("expected (0, 0), got (%d, %d)", pid, off)
	}
	got, err := sr.Fetch("greetings", 0, 0)
	if err != nil {
		t.Fatalf("Fetch: %v", err)
	}
	if string(got) != "hello" {
		t.Fatalf("expected 'hello', got %q", got)
	}
}

func TestNumPartitionsReflectsCreation(t *testing.T) {
	dir := t.TempDir()
	reg, _ := broker.Open(dir)
	sr := concurrent.New(reg, 100*time.Millisecond)
	defer sr.Close()

	sr.CreateTopic("t", 3)
	n, err := sr.NumPartitions("t")
	if err != nil {
		t.Fatalf("NumPartitions: %v", err)
	}
	if n != 3 {
		t.Fatalf("expected 3, got %d", n)
	}
}

func TestTopicNamesListsAllTopics(t *testing.T) {
	dir := t.TempDir()
	reg, _ := broker.Open(dir)
	sr := concurrent.New(reg, 100*time.Millisecond)
	defer sr.Close()

	sr.CreateTopic("beta", 1)
	sr.CreateTopic("alpha", 2)

	names := sr.TopicNames()
	if len(names) != 2 || names[0] != "alpha" || names[1] != "beta" {
		t.Fatalf("expected ['alpha', 'beta'], got %v", names)
	}
}

func TestProduceUnknownTopicErrors(t *testing.T) {
	dir := t.TempDir()
	reg, _ := broker.Open(dir)
	sr := concurrent.New(reg, 100*time.Millisecond)
	defer sr.Close()

	if _, _, err := sr.Produce("nope", []byte("x"), nil); err == nil {
		t.Fatal("expected error, got nil")
	}
}

func TestFetchUnknownTopicErrors(t *testing.T) {
	dir := t.TempDir()
	reg, _ := broker.Open(dir)
	sr := concurrent.New(reg, 100*time.Millisecond)
	defer sr.Close()

	if _, err := sr.Fetch("nope", 0, 0); err == nil {
		t.Fatal("expected error, got nil")
	}
}

// ── FetchBatch ────────────────────────────────────────────────────────────────

func TestFetchBatchReturnsUpToMaxCount(t *testing.T) {
	sr := makeShared(t, []struct{ name string; parts uint32 }{{"batch", 1}})
	for i := uint32(0); i < 50; i++ {
		b := make([]byte, 4)
		b[0] = byte(i >> 24); b[1] = byte(i >> 16); b[2] = byte(i >> 8); b[3] = byte(i)
		sr.Produce("batch", b, nil)
	}
	batch, err := sr.FetchBatch("batch", 0, 10, 20)
	if err != nil {
		t.Fatalf("FetchBatch: %v", err)
	}
	if len(batch) != 20 {
		t.Fatalf("expected 20 records, got %d", len(batch))
	}
	if batch[0].Offset != 10 {
		t.Fatalf("expected offset 10, got %d", batch[0].Offset)
	}
	if batch[19].Offset != 29 {
		t.Fatalf("expected offset 29, got %d", batch[19].Offset)
	}
}

func TestFetchBatchAtEndReturnsFewer(t *testing.T) {
	sr := makeShared(t, []struct{ name string; parts uint32 }{{"few", 1}})
	for i := uint32(0); i < 5; i++ {
		b := make([]byte, 4)
		b[3] = byte(i)
		sr.Produce("few", b, nil)
	}
	batch, err := sr.FetchBatch("few", 0, 3, 100)
	if err != nil {
		t.Fatalf("FetchBatch: %v", err)
	}
	if len(batch) != 2 {
		t.Fatalf("expected 2 records, got %d", len(batch))
	}
}

func TestFetchBatchEmptyTopic(t *testing.T) {
	sr := makeShared(t, []struct{ name string; parts uint32 }{{"empty", 1}})
	batch, err := sr.FetchBatch("empty", 0, 0, 10)
	if err != nil {
		t.Fatalf("FetchBatch: %v", err)
	}
	if len(batch) != 0 {
		t.Fatalf("expected empty, got %d records", len(batch))
	}
}

// ── concurrent produce ────────────────────────────────────────────────────────

func TestConcurrentProducesSinglePartition(t *testing.T) {
	sr := makeShared(t, []struct{ name string; parts uint32 }{{"events", 1}})

	var wg sync.WaitGroup
	for i := uint32(0); i < 4; i++ {
		wg.Add(1)
		go func(i uint32) {
			defer wg.Done()
			for j := uint32(0); j < 50; j++ {
				payload := []byte{byte(i), byte(j)}
				key := []byte{byte(i >> 24), byte(i >> 16), byte(i >> 8), byte(i)}
				if _, _, err := sr.Produce("events", payload, key); err != nil {
					t.Errorf("Produce: %v", err)
				}
			}
		}(i)
	}
	wg.Wait()

	// 4 threads × 50 = 200 messages, all hashed to partition 0 (only 1 partition).
	off, err := sr.NextOffset("events", 0)
	if err != nil {
		t.Fatalf("NextOffset: %v", err)
	}
	if off != 200 {
		t.Fatalf("expected 200, got %d", off)
	}
}

func TestConcurrentProducesSpreadAcrossPartitions(t *testing.T) {
	sr := makeShared(t, []struct{ name string; parts uint32 }{{"orders", 4}})

	var wg sync.WaitGroup
	for i := uint32(0); i < 8; i++ {
		wg.Add(1)
		go func(i uint32) {
			defer wg.Done()
			for j := uint32(0); j < 25; j++ {
				payload := []byte{byte(i), byte(j)}
				if _, _, err := sr.Produce("orders", payload, nil); err != nil {
					t.Errorf("Produce: %v", err)
				}
			}
		}(i)
	}
	wg.Wait()

	// 8 × 25 = 200 total across 4 partitions.
	total := totalOffsets(t, sr, "orders", 4)
	if total != 200 {
		t.Fatalf("expected 200 total, got %d", total)
	}
}

func TestConcurrentProduceAndFetchAfterJoin(t *testing.T) {
	sr := makeShared(t, []struct{ name string; parts uint32 }{{"log", 1}})

	producer := make(chan struct{})
	go func() {
		defer close(producer)
		for i := uint32(0); i < 100; i++ {
			b := make([]byte, 4)
			b[0] = byte(i >> 24); b[1] = byte(i >> 16); b[2] = byte(i >> 8); b[3] = byte(i)
			if _, _, err := sr.Produce("log", b, nil); err != nil {
				t.Errorf("Produce: %v", err)
			}
		}
	}()
	<-producer

	// After the producer finishes, all 100 messages must be fetchable in order.
	for i := uint64(0); i < 100; i++ {
		payload, err := sr.Fetch("log", 0, i)
		if err != nil {
			t.Fatalf("Fetch(%d): %v", i, err)
		}
		val := uint32(payload[0])<<24 | uint32(payload[1])<<16 | uint32(payload[2])<<8 | uint32(payload[3])
		if uint64(val) != i {
			t.Fatalf("message at offset %d has wrong value %d", i, val)
		}
	}
}

func TestConcurrentReadersWhileProducing(t *testing.T) {
	sr := makeShared(t, []struct{ name string; parts uint32 }{{"mixed", 1}})

	// Pre-write 50 records.
	for i := uint32(0); i < 50; i++ {
		b := make([]byte, 4)
		b[0] = byte(i >> 24); b[1] = byte(i >> 16); b[2] = byte(i >> 8); b[3] = byte(i)
		sr.Produce("mixed", b, nil)
	}

	writerDone := make(chan struct{})
	go func() {
		defer close(writerDone)
		for i := uint32(50); i < 150; i++ {
			b := make([]byte, 4)
			b[0] = byte(i >> 24); b[1] = byte(i >> 16); b[2] = byte(i >> 8); b[3] = byte(i)
			sr.Produce("mixed", b, nil)
		}
	}()

	var readerWg sync.WaitGroup
	for r := 0; r < 2; r++ {
		readerWg.Add(1)
		go func() {
			defer readerWg.Done()
			batch, err := sr.FetchBatch("mixed", 0, 0, 50)
			if err != nil {
				t.Errorf("FetchBatch: %v", err)
				return
			}
			for _, rec := range batch {
				val := uint32(rec.Payload[0])<<24 | uint32(rec.Payload[1])<<16 | uint32(rec.Payload[2])<<8 | uint32(rec.Payload[3])
				if uint64(val) != rec.Offset {
					t.Errorf("offset %d has wrong value %d", rec.Offset, val)
				}
			}
		}()
	}

	<-writerDone
	readerWg.Wait()

	off, _ := sr.NextOffset("mixed", 0)
	if off != 150 {
		t.Fatalf("expected 150, got %d", off)
	}
}

// ── background flush goroutine ────────────────────────────────────────────────

func TestFlushGoroutineDoesNotDeadlock(t *testing.T) {
	sr := makeShared(t, []struct{ name string; parts uint32 }{{"t", 1}})

	for i := uint32(0); i < 20; i++ {
		b := make([]byte, 4)
		b[3] = byte(i)
		sr.Produce("t", b, nil)
	}
	// Sleep past two flush intervals; the flush goroutine must not deadlock.
	time.Sleep(150 * time.Millisecond)
	off, err := sr.NextOffset("t", 0)
	if err != nil {
		t.Fatalf("NextOffset: %v", err)
	}
	if off != 20 {
		t.Fatalf("expected 20, got %d", off)
	}
}

func TestShutdownJoinsFlushGoroutineWithoutHanging(t *testing.T) {
	dir := t.TempDir()
	reg, _ := broker.Open(dir)
	sr := concurrent.New(reg, 50*time.Millisecond)
	sr.CreateTopic("x", 1)
	sr.Produce("x", []byte("data"), nil)
	sr.Close() // must return promptly
}

func TestExplicitFlushAllVisibleAfter(t *testing.T) {
	sr := makeShared(t, []struct{ name string; parts uint32 }{{"f", 1}})
	sr.Produce("f", []byte("msg"), nil)
	if err := sr.FlushAll(); err != nil {
		t.Fatalf("FlushAll: %v", err)
	}
	got, err := sr.Fetch("f", 0, 0)
	if err != nil {
		t.Fatalf("Fetch: %v", err)
	}
	if string(got) != "msg" {
		t.Fatalf("expected 'msg', got %q", got)
	}
}

File: rust/capstone-message-queue/tests/concurrent_test.rs
use mini_mq::broker::Registry;
use mini_mq::concurrent::SharedRegistry;
use std::sync::Arc;
use std::time::Duration;
use tempfile::TempDir;

// ── helpers ───────────────────────────────────────────────────────────────────

fn make_shared(dir: &TempDir, topics: &[(&str, u32)]) -> Arc<SharedRegistry> {
    let mut reg = Registry::open(dir.path()).unwrap();
    for &(name, parts) in topics {
        reg.create_topic(name, parts).unwrap();
    }
    Arc::new(SharedRegistry::new(reg, Duration::from_millis(50)))
}

fn total_offsets(shared: &SharedRegistry, topic: &str, num_partitions: u32) -> u64 {
    (0..num_partitions)
        .map(|p| shared.next_offset(topic, p).unwrap())
        .sum()
}

// ── single-threaded API sanity ────────────────────────────────────────────────

#[test]
fn create_topic_and_produce_fetch() {
    let dir = TempDir::new().unwrap();
    let shared = SharedRegistry::new(
        Registry::open(dir.path()).unwrap(),
        Duration::from_millis(100),
    );
    shared.create_topic("greetings", 1).unwrap();
    let (pid, off) = shared.produce("greetings", b"hello", None).unwrap();
    assert_eq!(pid, 0);
    assert_eq!(off, 0);
    assert_eq!(shared.fetch("greetings", 0, 0).unwrap(), b"hello");
}

#[test]
fn num_partitions_reflects_creation() {
    let dir = TempDir::new().unwrap();
    let shared = SharedRegistry::new(
        Registry::open(dir.path()).unwrap(),
        Duration::from_millis(100),
    );
    shared.create_topic("t", 3).unwrap();
    assert_eq!(shared.num_partitions("t").unwrap(), 3);
}

#[test]
fn topic_names_lists_all_topics() {
    let dir = TempDir::new().unwrap();
    let shared = SharedRegistry::new(
        Registry::open(dir.path()).unwrap(),
        Duration::from_millis(100),
    );
    shared.create_topic("beta", 1).unwrap();
    shared.create_topic("alpha", 2).unwrap();
    let names = shared.topic_names().unwrap();
    assert_eq!(names, vec!["alpha", "beta"]);
}

#[test]
fn produce_unknown_topic_errors() {
    let dir = TempDir::new().unwrap();
    let shared = SharedRegistry::new(
        Registry::open(dir.path()).unwrap(),
        Duration::from_millis(100),
    );
    assert!(shared.produce("nope", b"x", None).is_err());
}

#[test]
fn fetch_unknown_topic_errors() {
    let dir = TempDir::new().unwrap();
    let shared = SharedRegistry::new(
        Registry::open(dir.path()).unwrap(),
        Duration::from_millis(100),
    );
    assert!(shared.fetch("nope", 0, 0).is_err());
}

// ── fetch_batch ───────────────────────────────────────────────────────────────

#[test]
fn fetch_batch_returns_up_to_max_count() {
    let dir = TempDir::new().unwrap();
    let shared = make_shared(&dir, &[("batch", 1)]);
    for i in 0..50_u32 {
        shared.produce("batch", &i.to_be_bytes(), None).unwrap();
    }
    let batch = shared.fetch_batch("batch", 0, 10, 20).unwrap();
    assert_eq!(batch.len(), 20);
    assert_eq!(batch[0].offset, 10);
    assert_eq!(batch[19].offset, 29);
}

#[test]
fn fetch_batch_at_end_returns_fewer_than_max() {
    let dir = TempDir::new().unwrap();
    let shared = make_shared(&dir, &[("few", 1)]);
    for i in 0..5_u32 {
        shared.produce("few", &i.to_be_bytes(), None).unwrap();
    }
    let batch = shared.fetch_batch("few", 0, 3, 100).unwrap();
    assert_eq!(batch.len(), 2); // only offsets 3 and 4 exist
}

#[test]
fn fetch_batch_empty_topic() {
    let dir = TempDir::new().unwrap();
    let shared = make_shared(&dir, &[("empty", 1)]);
    let batch = shared.fetch_batch("empty", 0, 0, 10).unwrap();
    assert!(batch.is_empty());
}

// ── concurrent produce ────────────────────────────────────────────────────────

#[test]
fn concurrent_produces_single_partition() {
    let dir = TempDir::new().unwrap();
    let shared = make_shared(&dir, &[("events", 1)]);

    let mut handles = vec![];
    for i in 0..4_u32 {
        let s = Arc::clone(&shared);
        handles.push(std::thread::spawn(move || {
            for j in 0..50_u32 {
                let payload = format!("{i}-{j}");
                s.produce("events", payload.as_bytes(), Some(&i.to_be_bytes()))
                    .unwrap();
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }

    // 4 threads × 50 = 200 messages, all hashed to partition 0 (only 1 partition).
    assert_eq!(shared.next_offset("events", 0).unwrap(), 200);
}

#[test]
fn concurrent_produces_spread_across_partitions() {
    let dir = TempDir::new().unwrap();
    let shared = make_shared(&dir, &[("orders", 4)]);

    let mut handles = vec![];
    for i in 0..8_u32 {
        let s = Arc::clone(&shared);
        handles.push(std::thread::spawn(move || {
            for j in 0..25_u32 {
                let payload = format!("{i}-{j}");
                s.produce("orders", payload.as_bytes(), None).unwrap();
            }
        }));
    }
    for h in handles {
        h.join().unwrap();
    }

    // 8 × 25 = 200 total across 4 partitions.
    assert_eq!(total_offsets(&shared, "orders", 4), 200);
}

#[test]
fn concurrent_produce_and_fetch_after_join() {
    let dir = TempDir::new().unwrap();
    let shared = make_shared(&dir, &[("log", 1)]);

    let s = Arc::clone(&shared);
    let producer = std::thread::spawn(move || {
        for i in 0..100_u32 {
            s.produce("log", &i.to_be_bytes(), None).unwrap();
        }
    });
    producer.join().unwrap();

    // After the producer finishes, all 100 messages must be fetchable in order.
    for i in 0..100_u64 {
        let payload = shared.fetch("log", 0, i).unwrap();
        assert_eq!(
            u32::from_be_bytes(payload.try_into().unwrap()),
            i as u32,
            "message at offset {i} has wrong value"
        );
    }
}

#[test]
fn concurrent_readers_while_producing() {
    // Writer thread and two reader threads run simultaneously.
    // After the writer finishes, readers verify the data they saw was consistent
    // (no torn reads — each payload is either absent or intact).
    let dir = TempDir::new().unwrap();
    let shared = make_shared(&dir, &[("mixed", 1)]);

    // Pre-write 50 records so readers have something to start from.
    for i in 0..50_u32 {
        shared.produce("mixed", &i.to_be_bytes(), None).unwrap();
    }

    let s_write = Arc::clone(&shared);
    let writer = std::thread::spawn(move || {
        for i in 50..150_u32 {
            s_write.produce("mixed", &i.to_be_bytes(), None).unwrap();
        }
    });

    // Two readers that fetch batches while the writer runs.
    let mut reader_handles = vec![];
    for _ in 0..2 {
        let s = Arc::clone(&shared);
        reader_handles.push(std::thread::spawn(move || {
            let batch = s.fetch_batch("mixed", 0, 0, 50).unwrap();
            // Every record we DID see must have the right value.
            for rec in &batch {
                let val = u32::from_be_bytes(rec.payload.clone().try_into().unwrap());
                assert_eq!(val, rec.offset as u32);
            }
        }));
    }

    writer.join().unwrap();
    for h in reader_handles {
        h.join().unwrap();
    }

    assert_eq!(shared.next_offset("mixed", 0).unwrap(), 150);
}

// ── background flush thread ───────────────────────────────────────────────────

#[test]
fn flush_thread_does_not_deadlock() {
    let dir = TempDir::new().unwrap();
    let shared = make_shared(&dir, &[("t", 1)]);

    for i in 0..20_u32 {
        shared.produce("t", &i.to_be_bytes(), None).unwrap();
    }
    // Sleep past two flush intervals; the flush thread must not deadlock.
    std::thread::sleep(Duration::from_millis(150));
    // Still readable after flush thread ran.
    assert_eq!(shared.next_offset("t", 0).unwrap(), 20);
}

#[test]
fn shutdown_joins_flush_thread_without_hanging() {
    let dir = TempDir::new().unwrap();
    let reg = Registry::open(dir.path()).unwrap();
    // Drop triggers shutdown; must not hang.
    let shared = SharedRegistry::new(reg, Duration::from_millis(50));
    shared.create_topic("x", 1).unwrap();
    shared.produce("x", b"data", None).unwrap();
    drop(shared); // ← must return promptly
}

#[test]
fn explicit_flush_all_visible_after() {
    let dir = TempDir::new().unwrap();
    let shared = make_shared(&dir, &[("f", 1)]);
    shared.produce("f", b"msg", None).unwrap();
    shared.flush_all().unwrap();
    assert_eq!(shared.fetch("f", 0, 0).unwrap(), b"msg");
}
